### PR TITLE
refactor(menu): refactor component - [FE-2816] - [FE-2572] - [FE-2844]

### PR DESCRIPTION
**Pull request by dawidzarzycki**
In this commit `<Menu />` component will be updated to use Styled-Components.
This will allow us to use colors provided by theme object.
Because of the changes `<Menu />` component will be able to use colors provided
by app theme.
This commit also fixes issue when applying `onClick` prop to `<MenuItem />`
was changing its styles.

BREAKING CHANGE:
prop `as` has been replaced with `menuType`. Keyword `as` is reserved for
Styled-Components.
To update your components you will need to replace
`<Menu as="primary" />` with `<Menu menuType="primary" />`

### Proposed behaviour
new-menu
new-menu-submenu
new-menu-dark-submenu

### Current behaviour
menu-item-hover
submenu-item-hover

### Checklist

- [x] Screenshots are included in the PR
- [x] Carbon implementation and Design System documentation are congruent
- [x] All themes are supported
- [x] Commits follow our style guide
- [x] Unit tests added or updated
- [x] Cypress automation tests added or updated
- [ ] Storybook added or updated
- [ ] Typescript `d.ts` file added or updated

### Testing instructions
Storybook/Menu

Fixes: #2973
Fixes: #2903

### CODE REVIEW

**Comment by ianforster-sage** [comment]
Commit message for `refactor(menu): refactor component` has a missing quote
`<Menu as="primary />` with `<Menu menuType="primary" />`
should be this
`<Menu as="primary" />` with `<Menu menuType="primary" />`

File: src/components/menu/menu-item/menu-item.component.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+MenuItem.propTypes = {
+  /** Children elements */
+  children: PropTypes.node.isRequired,
+  /** Custom className */
+  className: PropTypes.string,
+  /** onClick handler */
+  onClick: PropTypes.func,
+  /** Adds an icon to the menu item. */
+  icon: PropTypes.string,
```
Comment (ianforster-sage) [inline src/components/menu/menu-item/menu-item.component.js:80]: Is this correctly defined? Shouldn't it be a type of `Icon`?

**Comment by dawidzarzycki** [comment]
> Commit message for `refactor(menu): refactor component` has a missing quote
> `<Menu as="primary />` with `<Menu menuType="primary" />`
> should be this
> `<Menu as="primary" />` with `<Menu menuType="primary" />`

good point! fixed

File: src/components/menu/menu-item/menu-item.component.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import Link from '../../link';
+import StyledMenuItemWrapper from './menu-item.style';
+import { StyledSubmenu, StyledSubmenuItem, StyledSubmenuTitle } from '../submenu-block/submenu.style';
+
+const MenuItem = (props) => {
+  const content = () => {
+    // if does not have a submenu, just render the children
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.component.js:10]: I think that it's already pretty self explanatory :)

File: src/components/menu/menu-item/menu-item.component.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import Link from '../../link';
+import StyledMenuItemWrapper from './menu-item.style';
+import { StyledSubmenu, StyledSubmenuItem, StyledSubmenuTitle } from '../submenu-block/submenu.style';
+
+const MenuItem = (props) => {
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.component.js:8]: I think that you could destructure these props

File: src/components/menu/menu-item/menu-item.style.js [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+    &:after {
+      transition: all 200ms;
+      content: "";
+      position: absolute;
+      height: 0;
+      background-color: transparent;
+      bottom: 0;
+      left: 0%;
+      right: 0%;
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.style.js:44]: do we need `%` here?

File: src/components/menu/menu-item/menu-item.style.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    &:hover:after {
+      background-color: ${theme.colors.primary}; 
+      height: 3px;
+      left: 0% !important;
+      right: 0% !important;
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.style.js:51]: why do we need `!important` here?

File: src/components/menu/menu-item/menu-item.style.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+  .carbon-portrait {
+    top: -1px;
+  }
+
+  &:focus {
+    z-index: 1;
+  }
+
+  ${({ menuType, theme }) => menuType === 'primary' && css`
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.style.js:22]: you could put that on top of the styled component:
`$({ menuType, theme, divide, hasSubMenu }) => css`
and then use it as just:
`$(menuType === 'primary' && css`

File: src/components/menu/menu-item/menu-item.style.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+      `}
+
+      ${menuType === 'secondary' && css`
+          content: "";
+          width: 0;
+          height: 0;
+          border-top: 5px solid ${theme.colors.white};
+          border-right: 4px solid transparent;
+          border-bottom: 4px solid transparent;
+          border-left: 4px solid transparent;
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.style.js:118]: if it differs only in border-top color, then you could just do:
`border-top: 5px solid ${menuType === 'primary' ? theme.colors. slate : theme.colors.white };`

File: src/components/menu/menu-item/menu-item.style.js [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            }
+          }
+
+          a {
+            text-decoration: none;
+
+            &:hover {
+              color: ${theme.colors.white};
+            }
+          }
```
Comment (grabkowski) [inline src/components/menu/menu-item/menu-item.style.js:151]:
```suggestion
          &:hover,
          &:hover a,
          a &:hover {
            color: ${theme.colors.white};
          }

          a {
            text-decoration: none;
          }
```

**Comment by DlgSHi** [comment]
onClick() issue failed QA

**Pull request merged**